codalab/apps/api/views/storage_views.py

```diff
@@ -0,0 +1,116 @@
+        }
+        resp_status = status.HTTP_200_OK
+        return Response(data, status=resp_status)
+
+
+@permission_classes((permissions.IsAuthenticated,))
+class GetCompetitionStorageAnalytics(views.APIView):
+    """
+    Gets storage use for competitions
+    """
+    def get(self, request, *args, **kwargs):
+        if not self.request.user.is_staff:
+            raise PermissionDenied(detail="Admin only")
+
+        comp_list = []
+        qs = Competition.objects.all().reverse()
+        for comp in qs:
+            comp_list.append(get_competition_size_data(comp))
+
+        return Response(comp_list, status=status.HTTP_200_OK)
+
+
+@permission_classes((permissions.IsAuthenticated,))
+class GetUserStorageAnalytics(views.APIView):
+    """
+    Gets storage use for users
+    """
+    def get(self, request, *args, **kwargs):
+        if not self.request.user.is_staff:
+            raise PermissionDenied(detail="Admin only")
+
+        user_list = []
+        qs = ClUser.objects.order_by('?').exclude(id=-1)
+        for user in qs:
+            user_list.append(user.get_storage_use_data())
+
+        return Response(user_list, status=status.HTTP_200_OK)
+
+
+@permission_classes((permissions.IsAuthenticated,))
+class GetCompetitionStorageAnalyticsOverTime(views.APIView):
+    """
+    Gets storage use based off of timeframes. IE: Use at that current point in time.
+    """
+
+    unit_table = {
+        'hours': ('hours', 1, "%H:%M:%S"),
+        'days': ('days', 1, "%m/%d/%Y"),
+        'weeks': ('weeks', 1, "%m/%d/%Y"),
+        'months': ('weeks', 4, "%b"),
+        'years': ('days', 365, "%Y")
+    }
+
+    def get(self, request, *args, **kwargs):
+        if not self.request.user.is_staff:
+            raise PermissionDenied(detail="Admin only")
+
+        unit_input = self.request.QUERY_PARAMS.get('unit', 'years')
+        sample_points = int(self.request.QUERY_PARAMS.get('sample_points', 5))
+
+        if not self.unit_table.get(unit_input, None):
+            unit_input = 'years'
+
+        unit, factor, date_format = self.unit_table[unit_input]
+        data_list = []
+
+        for index in range(0, sample_points):
+            total = 0
+            var_date = datetime.datetime.now() - datetime.timedelta(**{unit: factor*index})
+            if settings.USE_AWS:
+                bucket = BundleStorage.bucket
+                for key in bucket:
+                    # Returned time is in this format: 2020-10-13T20:15:50.000Z
+                    last_modified_date = parse_datetime(key.last_modified)
+                    if last_modified_date.replace(tzinfo=None) <= var_date:
+                        total += key.size
+            else:
+                found_files = set(storage_recursive_find(BundleStorage))
+                for file_path in found_files:
+                    if BundleStorage.modified_time(file_path) <= var_date:
+                        total += BundleStorage.size(file_path) or 0
```

Comment:
is there a bulk way to get size of a list of files?

Reply:
It does not look like it for Boto or Django's storage classes.

For boto(2&3), every implementation I have seen revolves around getting a list of files, and still iterating over them manually to sum the size. I have not seen anything in the documentation either regarding this.

For Django storage classes, we're very limited on useful methods for this case:
https://docs.djangoproject.com/en/1.8/ref/files/storage/#the-storage-class

Reply:
you could make a tiny tiny flask server running on the Minio box that you could have an `/api/storage_analytics ` endpoint at. you'd need to pass some kind of secret key but this service could return JSON giving you storage details. Commands would run near-instantly and not DDoS the server.